integration/kotlinx-coroutines-play-services/test/TaskTest.kt

```diff
@@ -149,5 +149,277 @@ class TaskTest : TestBase() {
+        val completionException = deferred.getCompletionExceptionOrNull()!!
+        assertTrue(completionException is TestException)
+        assertEquals("something went wrong", completionException.message)
+
+        try {
+            deferred.await()
+            fail("deferred.await() should throw an exception")
+        } catch (e: Exception) {
+            assertTrue(e is TestException)
+            assertEquals("something went wrong", e.message)
+        }
+        assertFalse(cancellationTokenSource.token.isCancellationRequested)
+    }
+
+    @Test
+    fun testFailingCancellableTaskAsDeferred() = runTest {
+        val cancellationTokenSource = CancellationTokenSource()
+        val lock = ReentrantLock().apply { lock() }
+
+        val deferred: Deferred<Int> = Tasks.call {
+            lock.withLock { throw TestException("something went wrong") }
+        }.asDeferred(cancellationTokenSource)
+
+        assertFalse(deferred.isCompleted)
+        lock.unlock()
+
+        try {
+            deferred.await()
+            fail("deferred.await() should throw an exception")
+        } catch (e: Exception) {
+            assertTrue(e is TestException)
+            assertEquals("something went wrong", e.message)
+            assertSame(e.cause, deferred.getCompletionExceptionOrNull()) // debug mode stack augmentation
+        }
+        assertFalse(cancellationTokenSource.token.isCancellationRequested)
+    }
+
+    @Test
+    fun testFastPathCompletedTaskWithCancelledTokenSourceAsDeferred() = runTest {
+        val cancellationTokenSource = CancellationTokenSource()
+        val deferred = Tasks.forResult(42).asDeferred(cancellationTokenSource)
+        cancellationTokenSource.cancel()
+        assertEquals(42, deferred.await())
+    }
+
+    @Test
+    fun testAwaitCancellableTask() = runTest {
+        val cancellationTokenSource = CancellationTokenSource()
+        val taskCompletionSource = TaskCompletionSource<Int>(cancellationTokenSource.token)
+
+        val deferred: Deferred<Int> = async(start = CoroutineStart.UNDISPATCHED) {
+            taskCompletionSource.task.await(cancellationTokenSource)
+        }
+
+        assertFalse(deferred.isCompleted)
+        taskCompletionSource.setResult(42)
+
+        assertEquals(42, deferred.await())
+        assertTrue(deferred.isCompleted)
+    }
+
+    @Test
+    fun testFailedAwaitTask() = runTest(expected = { it is TestException }) {
+        val cancellationTokenSource = CancellationTokenSource()
+        val taskCompletionSource = TaskCompletionSource<Int>(cancellationTokenSource.token)
+
+        val deferred: Deferred<Int> = async(start = CoroutineStart.UNDISPATCHED) {
+            taskCompletionSource.task.await(cancellationTokenSource)
+        }
+
+        assertFalse(deferred.isCompleted)
+        taskCompletionSource.setException(TestException("something went wrong"))
+
+        deferred.await()
+    }
+
+    @Test
+    fun testCancelledAwaitCancellableTask() = runTest {
```

Comment:
This test seems to fail consistently on our CI: https://teamcity.jetbrains.com/buildConfiguration/KotlinTools_KotlinxCoroutines_BuildAggregated/3519632?expandedTest=build%3A%28id%3A3519630%29%2Cid%3A44777

PTAL

Reply:
I haven't been able to reproduce locally, but I think I narrowed down what's happening. The latest changes should pass!

The `Task` internals have some locks and also `post` something onto the main looper. For these tests, it looks like the looper is faked in `FakeAndroid`, which ends up doing a `GlobalScope.launch { ... }`, which I believe implies that the task won't synchronously know if it has been cancelled or not.

My solution right now is to just remove those assertions, as arguably checking that the task is actually cancelled is the `Task` library's responsibility. Since the main looper is faked, we could probably control the race, but that's probably not very helpful either since real usage wouldn't use a fake looper.